File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/socialmedia/DeleteReaction.java
package com.github.dedis.popstellar.model.network.method.message.data.socialmedia;

import com.github.dedis.popstellar.model.network.method.message.data.*;
import com.github.dedis.popstellar.model.objects.security.MessageID;
import com.github.dedis.popstellar.utility.MessageValidator;
import com.google.gson.annotations.SerializedName;

public class DeleteReaction extends Data {

  @SerializedName("reaction_id")
  private final MessageID reactionID;

  private final long timestamp;

  public DeleteReaction(MessageID reactionID, long timestamp) {
    MessageValidator.verify().isBase64(reactionID.getEncoded(), "reaction id");
    this.reactionID = reactionID;
    this.timestamp = timestamp;
  }

  @Override
  public String getObject() {
    return Objects.REACTION.getObject();
  }

  @Override
  public String getAction() {
    return Action.DELETE.getAction();
  }

  public MessageID getReactionID() {
    return reactionID;
  }

  public long getTimestamp() {
    return timestamp;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    DeleteReaction that = (DeleteReaction) o;
    return timestamp == that.timestamp && java.util.Objects.equals(reactionID, that.reactionID);
  }

  @Override
  public int hashCode() {
    return java.util.Objects.hash(reactionID, timestamp);
  }

  @Override
  public String toString() {
    return "DeleteReaction{"
        + "reactionID='"
        + reactionID
        + '\''
        + ", timestamp="
        + timestamp
        + '}';
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/socialmedia/DeleteReaction.kt
package com.github.dedis.popstellar.model.network.method.message.data.socialmedia

import com.github.dedis.popstellar.model.network.method.message.data.Action
import com.github.dedis.popstellar.model.network.method.message.data.Data
import com.github.dedis.popstellar.model.network.method.message.data.Objects
import com.github.dedis.popstellar.model.objects.security.MessageID
import com.github.dedis.popstellar.utility.MessageValidator.verify
import com.google.gson.annotations.SerializedName

class DeleteReaction(reactionID: MessageID, timestamp: Long) : Data {
  @SerializedName("reaction_id") val reactionID: MessageID
  val timestamp: Long

  init {
    verify().isNotEmptyBase64(reactionID.encoded, "reaction id").validPastTimes(timestamp)

    this.reactionID = reactionID
    this.timestamp = timestamp
  }

  override val `object`: String
    get() = Objects.REACTION.`object`

  override val action: String
    get() = Action.DELETE.action

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    val that = other as DeleteReaction
    return timestamp == that.timestamp && reactionID == that.reactionID
  }

  override fun hashCode(): Int {
    return java.util.Objects.hash(reactionID, timestamp)
  }

  override fun toString(): String {
    return "DeleteReaction{reactionID='$reactionID', timestamp=$timestamp}"
  }
}
